// tubular-react/dist/DataGrid/DataGrid.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var LinearProgress_1 = require("@material-ui/core/LinearProgress");
var Paper_1 = require("@material-ui/core/Paper");
var Table_1 = require("@material-ui/core/Table");
var TableHead_1 = require("@material-ui/core/TableHead");
var TableRow_1 = require("@material-ui/core/TableRow");
var makeStyles_1 = require("@material-ui/styles/makeStyles");
var React = require("react");
var tubular_react_common_1 = require("tubular-react-common");
var uno_react_1 = require("uno-react");
var Pagination_1 = require("../Pagination");
var GridToolbar_1 = require("../Toolbar/GridToolbar");
var ToolbarOptions_1 = require("../Toolbar/ToolbarOptions");
var _1 = require("./");
var MobileDataGridTable_1 = require("./MobileDataGridTable");
var useStyles = makeStyles_1.default({
    linearProgress: {
        marginTop: '-10px',
        height: '20px',
    },
    root: {
        overflowX: 'auto',
        width: '100%',
    },
});
var timeout = 400;
exports.DataGrid = function (props) {
    var columns = props.columns, dataSource = props.dataSource, deps = props.deps, footerComponent = props.footerComponent, gridName = props.gridName, _a = props.mobileBreakpointWidth, mobileBreakpointWidth = _a === void 0 ? props.mobileBreakpointWidth || 800 : _a, onError = props.onError, onRowClick = props.onRowClick, rowComponent = props.rowComponent, rowMobileComponent = props.rowMobileComponent, storage = props.storage, _b = props.toolbarOptions, toolbarOptions = _b === void 0 ? props.toolbarOptions || new ToolbarOptions_1.ToolbarOptions() : _b, detailComponent = props.detailComponent;
    var classes = useStyles({});
    var tbTableInstance = tubular_react_common_1.useTbTable(columns, dataSource, {
        callbacks: { onError: onError },
        componentName: gridName,
        deps: deps,
        pagination: {
            itemsPerPage: toolbarOptions.itemsPerPage,
        },
        storage: storage,
    });
    var isMobileResolution = uno_react_1.useResolutionSwitch(mobileBreakpointWidth, timeout)[0];
    if (isMobileResolution) {
        toolbarOptions.SetMobileMode();
        return (React.createElement(Paper_1.default, { className: classes.root },
            React.createElement(GridToolbar_1.GridToolbar, { toolbarOptions: toolbarOptions, tbTableInstance: tbTableInstance, gridName: gridName }),
            React.createElement("div", { className: classes.linearProgress, "data-testid": "linear-progress" }, tbTableInstance.state.isLoading && React.createElement(LinearProgress_1.default, null)),
            React.createElement(MobileDataGridTable_1.MobileDataGridTable, { tbTableInstance: tbTableInstance, onRowClick: onRowClick, rowComponent: rowMobileComponent }),
            React.createElement(Pagination_1.Paginator, { advancePagination: toolbarOptions.advancePagination, rowsPerPageOptions: toolbarOptions.rowsPerPageOptions, tbTableInstance: tbTableInstance })));
    }
    var paginator = function (position) { return (React.createElement(Table_1.default, { "data-testid": position + "-paginator" },
        React.createElement(TableHead_1.default, null,
            React.createElement(TableRow_1.default, null,
                React.createElement(Pagination_1.Paginator, { advancePagination: toolbarOptions.advancePagination, rowsPerPageOptions: toolbarOptions.rowsPerPageOptions, tbTableInstance: tbTableInstance }))))); };
    return (React.createElement(Paper_1.default, { className: classes.root },
        React.createElement(GridToolbar_1.GridToolbar, { gridName: gridName, toolbarOptions: toolbarOptions, tbTableInstance: tbTableInstance }),
        toolbarOptions.topPager && paginator('top'),
        React.createElement("div", { className: classes.linearProgress, "data-testid": "linear-progress" }, tbTableInstance.state.isLoading && React.createElement(LinearProgress_1.default, null)),
        React.createElement(_1.DataGridTable, { tbTableInstance: tbTableInstance, rowComponent: rowComponent, footerComponent: footerComponent, detailComponent: detailComponent || null, onRowClick: onRowClick }),
        toolbarOptions.bottomPager && paginator('bottom')));
};
